Cli for M365 release post v6.2 (#960)
pnp/blog: content/cli-for-microsoft-365/cli-for-microsoft-365-v6-2/index.md

CLI for Microsoft 365 is a cross-platform CLI that allows you to manage various configuration settings of Microsoft 365 and SharePoint Framework projects no matter which operating system or shell you use.

While building solutions for Microsoft 365 expands beyond the Windows operating system, managing many platform settings is possible only through PowerShell on Windows. As more and more users work on non-Windows machines, it's inconvenient for them to use a Windows virtual machine to configure their tenants. With the CLI for Microsoft 365, you can configure your tenant no matter which operating system you use. Additionally, using CLI for Microsoft 365, you can manage your SharePoint Framework projects.

## New version of CLI for Microsoft 365 – v6.2

Following our monthly release cadence, we've released a new version of the CLI for Microsoft 365 with some new capabilities. The month of January was no exception. We had a ton of improvements and new commands. Here are a few of the most noteworthy additions.

> For the complete list of what's new and changed, see the [release notes](https://pnp.github.io/cli-microsoft365/about/release-notes/#v620).

## Added support to share files

> SharePoint is all about collaborating with people on documents. Previously we introduced some commands to manage permissions on documents. With our new commands you can share your SharePoint files with anyone you like. In this release we are introducing commands to manage sharing links on documents.

Reviewer: ```suggestion
SharePoint is all about collaborating with people on documents. Previously we introduced some commands to manage permissions on documents. With our new commands, you can share your SharePoint files with anyone you like. In this release, we are introducing commands to manage sharing links on documents.
```

To create an anonymous sharing link for a file, run:

```sh
m365 spo file sharinglink add --webUrl https://contoso.sharepoint.com --fileUrl "/sites/demo/Shared Documents/Test1.docx" --type edit --scope anonymous --expirationDateTime "2023-01-09T16:20:00Z"
```

To list all anonymous sharing links of a file, run:

```sh
m365 spo file sharinglink list --webUrl https://contoso.sharepoint.com/sites/demo --fileUrl "/sites/demo/shared documents/document.docx" --scope anonymous
```

To remove a sharing link of a file, run:

```sh
m365 spo file sharinglink remove --webUrl https://contoso.sharepoint.com/sites/demo --fileUrl '/sites/demo/Shared Documents/document.docx' --id c391b57d-5783-4c53-9236-cefb5c6ef323
```

More sharing link commands coming soon!

## Application permissions for Planner

> Eureka! Finally we can use application permissions for Planner commands. This much requested feature is currently rolling out preview. Previously we showed a friendly error that told that application permissions don't work for Planner commands, this check has now been removed. You are now free to experiment with our Planner commands with application permissions! You can do this quickly and easy by just creating a new application registration in Azure AD with Planner application permissions, logging to the CLI with a [certificate](https://pnp.github.io/cli-microsoft365/user-guide/connecting-office-365/#log-in-using-a-certificate) or [secret](https://pnp.github.io/cli-microsoft365/user-guide/connecting-office-365/#log-in-using-a-secret), and you are good to go!

Reviewer 2: _is currently rolling out **in** preview_

> Eureka! Finally we can use application permissions for Planner commands. This much requested feature is currently rolling out in preview. Previously we showed a friendly error that told that application permissions don't work for Planner commands, this check has now been removed. You are now free to experiment with our Planner commands with application permissions! You can do this quickly and easy by just creating a new application registration in Azure AD with Planner application permissions, logging in to the CLI with a [certificate](https://pnp.github.io/cli-microsoft365/user-guide/connecting-office-365/#log-in-using-a-certificate) or [secret](https://pnp.github.io/cli-microsoft365/user-guide/connecting-office-365/#log-in-using-a-secret), and you are good to go!

Reviewer: ```suggestion
Eureka! Finally, we can use application permissions for Planner commands. This much-requested feature is currently rolling out in preview. Previously we showed a friendly error that told us that application permissions don't work for Planner commands, this check has now been removed. You are now free to experiment with our Planner commands with application permissions! You can do this quickly and easily by just creating a new application registration in Azure AD with Planner application permissions, logging in to the CLI with a [certificate](https://pnp.github.io/cli-microsoft365/user-guide/connecting-office-365/#log-in-using-a-certificate) or [secret](https://pnp.github.io/cli-microsoft365/user-guide/connecting-office-365/#log-in-using-a-secret), and you are good to go!
```

## More compliance commands

> In our previous release we introduced commands to manage compliance within your tenant. This was just the beginning, our v6.2 release contains even more compliance commands.

Reviewer: ```suggestion
In our previous release, we introduced commands to manage compliance within your tenant. This was just the beginning, our v6.2 release contains even more compliance commands.
```

Reviewer: Maybe we could provide a bit more insight here regarding all the new commands mentioned here

> To add create a new retention label, run:

Reviewer 3: To add create?

```sh
m365 purview retentionlabel add --displayName 'some label' --behaviorDuringRetentionPeriod retain --actionAfterRetentionPeriod delete --retentionDuration 365
```

> To add/set a retention label on a file, run:

Reviewer 3: We could also use the word ‘apply’ here as that’s what you do: apply a label. Add/set is CLI words

```sh
m365 spo file retentionlabel ensure --webUrl https://contoso.sharepoint.com/sites/project-x --fileUrl '/Shared Documents/Document.docx' --name 'Some label'
```

> To remove a retention label from a folder, run:

Reviewer 3: Use the word clear instead of remove is better

```sh
m365 spo folder retentionlabel remove --webUrl https://contoso.sharepoint.com/sites/project-x --folderId 0cd891ef-afce-4e55-b836-fce03286cccf
```

To lock a list item record, run:

```sh
m365 spo listitem record lock --webUrl https://contoso.sharepoint.com/sites/project-x --listId 0cd891ef-afce-4e55-b836-fce03286cccf --listItemId 1
```

> Checkout our [release notes](https://pnp.github.io/cli-microsoft365/about/release-notes/#v620) for more Purview related commands.

Reviewer: ```suggestion
Check out our [release notes](https://pnp.github.io/cli-microsoft365/about/release-notes/#v620) for more Purview related commands.
```

> Check out our [release notes](https://pnp.github.io/cli-microsoft365/about/release-notes/#v620) for more Purview related commands.

Reviewer 3: Maybe add a small section naming the additional commands would be good…

## Further steps in implementing CLI for Microsoft 365 context

> In our previous release we have taken the first steps to create a CLI context. This allows you to save time by storing default values for options that you often use. For example, if you're working with a specific site collection, you can save the URL of the site collection in the context, and CLI for Microsoft 365 will automatically use it when you run a command that requires the URL of the site collection. This way, you don't have to specify the URL of the site collection every time you run a command. In this release we created extra commands to work with the CLI context.

Reviewer 2: Let's clarify that this is the idea, but we don't support context just yet. We allow managing context settings but don't apply them to commands just yet.

> In our previous release we have taken the first steps to create a CLI context. This allows you to save time by storing default values for options that you often use. For example, if you're working with a specific site collection, you can save the URL of the site collection in the context, and CLI for Microsoft 365 will automatically use it when you run a command that requires the URL of the site collection. This way, you don't have to specify the URL of the site collection every time you run a command. In this release we created extra commands to work with the CLI context. Note this is still work in progress. We allow managing context settings but don't apply them to commands just yet.

Reviewer: ```suggestion
In our previous release, we have taken the first steps to create a CLI context. This allows you to save time by storing default values for options that you often use. For example, if you're working with a specific site collection, you can save the URL of the site collection in the context, and CLI for Microsoft 365 will automatically use it when you run a command that requires the URL of the site collection. This way, you don't have to specify the URL of the site collection every time you run a command. In this release, we created extra commands to work with the CLI context. Note this is still a work in progress. We allow managing context settings but don't apply them to commands just yet.
```

To add/set a context value for a certain option, run:

```sh
m365 context option set --name 'listName' --value 'testList'
```

To remove a context value, run:

```sh
m365 context option remove --name "listName"
```

## Extended Power Platform commands

> In the past releases we gave more focus to our Power Platform commands. This month is not different. Also, in this release we have new Power Platform commands, especially about AI Builder Models.

Reviewer: ```suggestion
In the past releases, we gave more focus to our Power Platform commands. This month is no different. Also, in this release, we have new Power Platform commands, especially about AI Builder Models.
```

To list all AI Builder Models in a specific environment, run:

```sh
m365 pp aibuildermodel list --environment "Default-d87a7535-dd31-4437-bfe1-95340acd55c5"
```

To remove an AI Builder Model, run:

```sh
m365 pp aibuildermodel remove --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --name "AI Builder Model Name"
```

> 

Reviewer 2: I suggest we add information about updates to booleans - which significantly simplify using the CLI, and CI/CD guidance to make it easier for folks to use CLI in their deployment pipelines.

Author: I saw that this was already included in our previous release post, or do you mean something else?

https://github.com/pnp/blog/blob/6727797d63148f03298f373e8258ed8eb50cbb01/content/cli-for-microsoft-365/cli-for-microsoft-365-v6-1/index.md?plain=1#L147-L151

Reviewer 2: Ah right, seems like in this release we only updated the guidance. Never mind. Carry on.

## What's next

Here are some things that we are currently working on.

### Planner Rosters

> Rosters are a new type of container for Microsoft Planner plans. This enables users to create a Planner plan without the need to create a new Microsoft 365 group (with a mailbox, SharePoint site, ...). Access to Roster-contained plans is controlled by the members on the Roster. However Planner Rosters are still in preview, we are already working on providing the necessary commands for this.

Reviewer: ```suggestion
Rosters are a new type of container for Microsoft Planner plans. This enables users to create a Planner plan without the need to create a new Microsoft 365 group (with a mailbox, SharePoint site, ...). Access to Roster-contained plans is controlled by the members of the Roster. However Planner Rosters are still in preview, we are already working on providing the necessary commands for this.
```

> Rosters are a new type of container for Microsoft Planner plans. This enables users to create a Planner plan without the need to create a new Microsoft 365 group (with a mailbox, SharePoint site, ...). Access to Roster-contained plans is controlled by the members of the Roster. However Planner Rosters are still in preview, we are already working on providing the necessary commands for this.

Reviewer 3: Even though planner rosters are still in preview…

### Sharing link commands

> With this release you are able to easily manage sharing links for SharePoint files. We don't stop here, we have more commands on our roadmap to manage sharing links. We are planning to create more commands to manage sharing links for SharePoint folders, SharePoint list items, OneDrive files, ...

Reviewer: ```suggestion
With this release, you are able to easily manage sharing links for SharePoint files. We don't stop here, we have more commands on our roadmap to manage sharing links. We are planning to create more commands to manage sharing links for SharePoint folders, SharePoint list items, OneDrive files, ...
```

### Manage SPFx extensions

> Currently CLI for Microsoft 365 support few commands for SPFx solutions. In the upcoming weeks we are planning to create more SPFx commands with a focus on SPFx extensions. Planning to add, set, list, remove SPFx extensions such as ListView Command Sets or Application Customizers? We provide everything to make your life easier.

Reviewer: ```suggestion
Currently, CLI for Microsoft 365 supports a few commands for SPFx solutions. In the upcoming weeks, we are planning to create more SPFx commands with a focus on SPFx extensions. Planning to add, set, list, and remove SPFx extensions such as ListView Command Sets or Application Customizers? We provide everything to make your life easier.
```

## Contributors

This release wouldn't be possible without the help of (in alphabetical order):

- [Adam Wójcik](https://github.com/Adam-it)
- [Arjun Menon](https://github.com/arjunumenon)
- [Jasey Waegebaert](https://github.com/Jwaegebaert)
- [Martin Lingstuyl](https://github.com/martinlingstuyl)
- [Mathijs Verbeeck](https://github.com/MathijsVerbeeck)
- [Michael Maillot](https://github.com/michaelmaillot)
- [Milan Holemans](https://github.com/milanholemans)
- [Nanddeep Nachan](https://github.com/nanddeepn)
- [Nico De Cleyre](https://github.com/nicodecleyre)
- [Patrick Lamber](https://github.com/plamber)
- [Smita Nachan](https://github.com/SmitaNachan)

### New contributors

A warm welcome to this month's new contributors! (in alphabetical order):

> - [Tim Van de Voorde](https://github.com/timdelavoorde)

Reviewer 2: I haven't seen any PRs from Tim in the git story, or have I missed it?

Author: He made a PR https://github.com/pnp/cli-microsoft365/pull/4271, but something went wrong while committing his PR, his name was dropped of the commit. https://github.com/pnp/cli-microsoft365/commit/58ae9e6b56f5f2d38c44e93dbf3d9ca26cdc7a90

Reviewer 2: Did we somehow lost him when merging his PR? That's a shame 😞 If it's not too far back, we could see if we could update the PRs author to him to attribute his work to him.

Author: That will require a force push on the main branch right?

Reviewer 2: Correct
